posthog/async_migrations/migrations/0004_replicated_schema.py

```diff
@@ -143,26 +158,33 @@ def replicated_table_operations(self, table: TableMigrationData):
 
         yield AsyncMigrationOperation(
             fn=lambda _: self.rename_tables(
-                table.name, table.tmp_table_name, table.renamed_table_name, table.backup_table_name
+                [table.name, table.backup_table_name], [table.tmp_table_name, table.renamed_table_name],
             ),
             rollback_fn=lambda _: self.rename_tables(
-                table.backup_table_name,
-                table.renamed_table_name,
-                table.tmp_table_name,
-                table.name,
-                skip_if_backup_exists=True,
+                [table.renamed_table_name, table.tmp_table_name],
+                [table.backup_table_name, table.name],
+                verify_table_exists=table.backup_table_name,
             ),
         )
 
-        if table.materialized_view_name is not None:
-            yield AsyncMigrationOperation.simple_op(
-                sql=table.create_materialized_view, rollback=f"DROP TABLE IF EXISTS {table.materialized_view_name}",
-            )
-
+    def finalize_table_operations(self, table: TableMigrationData):
         # NOTE: Relies on IF NOT EXISTS on the query
         if isinstance(table, ShardedTableMigrationData):
-            for create_table_query in table.extra_tables:
-                yield AsyncMigrationOperation.simple_op(sql=create_table_query)
+            for table_name, create_table_query in table.extra_tables.items():
+                yield AsyncMigrationOperationSQL(sql=create_table_query, rollback=f"DROP TABLE IF EXISTS {table_name}")
+
+        if isinstance(table, ShardedTableMigrationData) and table.materialized_view_name is not None:
```

Comment:
I believe this lingered? see line 184 - it's the same condition + I believe it's the correct one

Reply:
Nope. This is needed for this whole thing to work.

Problem is that materialized views _can_ exist without kafka tables being there, but _cannot_ be created. All of this is a hack to ensure things are done in the right sequence even in a rollback.

Reply:
Right, supposed this could have been the case 👍